datumaro/components/config_model.py

```diff
@@ -27,35 +34,108 @@ def __init__(self, config=None):
         super().__init__(config, schema=MODEL_SCHEMA)
 
 
+BUILDSTAGE_SCHEMA = _SchemaBuilder() \
+    .add('name', str) \
+    .add('type', str) \
+    .add('kind', str) \
+    .add('hash', str) \
+    .add('params', dict) \
+    .build()
+
+class BuildStage(Config):
+    def __init__(self, config=None):
+        super().__init__(config, schema=BUILDSTAGE_SCHEMA)
+
+BUILDTARGET_SCHEMA = _SchemaBuilder() \
+    .add('stages', list) \
+    .add('parents', list) \
+    .build()
+
+class BuildTarget(Config):
+    def __init__(self, config=None):
+        super().__init__(config, schema=BUILDTARGET_SCHEMA)
+        self.stages = [BuildStage(o) for o in self.stages]
+
+    @property
+    def root(self):
+        return self.stages[0]
+
+    @property
+    def head(self):
+        return self.stages[-1]
+
+    @property
+    def has_stages(self) -> bool:
+        return len(self.stages) == 1
```

Comment:
The reason is that the first stage (called `root`) is the unmodified source. It should always be present. It could be used in pipeline building, but it looks like now it is done another way.

Reply:
> The reason is that the first stage (called root) is the unmodified source. It should always be present.

That doesn't explain why a function called `has_stages` checks if there's exactly one stage. If `root` should always be present, I would expect that to be handled by an assert or something like that.

> It could be used in pipeline building, but it looks like now it is done another way.

Then perhaps it should just be removed?

Reply:
I don't agree that all the currently unused code must be removed.

Reply:
What about the first half of my comment?

Reply:
Because it is an error. Fixed.

Reply:
You kinda made the case for removing unused code for me there. :P

Oh well. At least the error is fixed.